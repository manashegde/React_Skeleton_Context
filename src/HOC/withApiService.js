import React from "react";

const xbcGatewayServicesElem = document.getElementById("xbcGatewayServicesEp");
const xbcGatewayUrl = xbcGatewayServicesElem
  ? xbcGatewayServicesElem.getAttribute("di")
  : "/nic-xbc-gateway/xbc/gateway/";

const nicIdentityServiceElem = document.getElementById("nicIdentityService");
const nicIdentityServiceUrl = nicIdentityServiceElem
  ? nicIdentityServiceElem.getAttribute("di")
  : "http://10.0.1.6:50010/mocks-gateway/xbc/mocks/gateway/";

export const urls = {
  login: `${nicIdentityServiceUrl}login/v1`,
  refreshToken: `${nicIdentityServiceUrl}token/refresh/v1`,
  postLogin: `${xbcGatewayUrl}security/post-login`,

  crossing: {
    arrival: {
      search: {
        citizen: `${xbcGatewayUrl}crossing/search/arrival/citizen`,
        alien: `${xbcGatewayUrl}crossing/search/arrival/alien`
      },
      complete: {
        citizen: `${xbcGatewayUrl}crossing/arrival/complete/citizen`,
        alien: `${xbcGatewayUrl}crossing/arrival/complete/alien`
      }
    },
    departure: {
      search: {
        citizen: `${xbcGatewayUrl}crossing/search/departure/citizen`
      },
      complete: {
        citizen: `${xbcGatewayUrl}crossing/departure/complete/citizen`
      }
    },
    specialSearch: `${xbcGatewayUrl}crossing/search/special`,
    lookup: `${xbcGatewayUrl}lookup/ref-data/`, //Bcr269BusRule/readAll`,
    fetchTripInfo: `${xbcGatewayUrl}crossing/search/tripInfo/fetch`,
    updateTripInfo: `${xbcGatewayUrl}crossing/search/tripInfo/update`,
    searchCwl: `${xbcGatewayUrl}crossing/cwl/search`,
    updateCwl: `${xbcGatewayUrl}crossing/cwl/update`,
    searchCwlDetails: `${xbcGatewayUrl}crossing/cwl/details`
  }
};

export function makeLookupUrl({ entity, pk, isReadAll }) {
  let out = urls.crossing.lookup + entity;
  if (isReadAll) {
    return out + "/readAll";
  } else {
    return out + "/read" + (pk ? `/${pk}` : "");
  }
}

export function makeCrossingUrlFor(action, travelerType, crossingType) {
  try {
    const { crossing } = urls;
    const result = [crossingType, action, travelerType].reduce((prev, curr) => {
      return prev[curr];
    }, crossing);
    console.log("------crossing url:", result);
    return result;
  } catch (e) {
    throw new Error(
      "failed to make crossing url: missing action, travelerType, or crossingType"
    );
  }
}

export function parseJwt(token) {
  var base64Url = token.split(".")[1];
  var base64 = base64Url.replace(/-/g, "+").replace(/_/g, "/");
  return JSON.parse(window.atob(base64));
}

export function post({ url, data, method }) {
  const jwt2 = sessionStorage.getItem("jwt2") || "";
  let headers = {
    "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8"
  };
  if (jwt2) headers["Authorization"] = `Bearer ${jwt2}`;
  return fetch(url, {
    body: JSON.stringify(data), // must match 'Content-Type' header
    headers,
    method: method || "POST", // *GET, POST, PUT, DELETE, etc.
    referrer: "no-referrer" // *client, no-referrer
  })
    .then(res => res.json().catch(err => res))
    .then(res => {
      if (res.status && res.status !== 200) {
        throw res;
      } else {
        return res;
      }
    })
    .catch(e => {
      throw e.message || e.statusText || "unknown post error";
    });
}

export function get(url) {
  return fetch(url)
    .then(res => res.json())
    .catch(e => {
      console.error("--------api get:", e);
      throw e.message || e.statusText || "unknown get error";
    });
}

//DO NOT USE THIS ANYMORE.
//USE POST AND GET METHODS FROM SESSIONCONTEXT INSTEAD
export default function withApiService(WrappedComponent) {
  class C extends React.Component {
    render() {
      return (
        <WrappedComponent
          {...this.props}
          apiService={{ post, get, urls, parseJwt }}
        />
      );
    }
  }
  C.displayName = `withApiService(${WrappedComponent.displayName ||
    WrappedComponent.name})`;
  return C;
}
