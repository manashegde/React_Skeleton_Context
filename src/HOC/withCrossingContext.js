import React from "react";
import { CrossingContextConsumer } from "contexts/crossing";

export default function withCrossingContext(WrappedComponent) {
  class C extends React.Component {
    render() {
      return (
        <CrossingContextConsumer>
          {crossingContext => (
            <WrappedComponent
              {...this.props}
              crossingContext={crossingContext}
            />
          )}
        </CrossingContextConsumer>
      );
    }
  }
  C.displayName = `withCrossingContext(${WrappedComponent.displayName ||
    WrappedComponent.name})`;
  return C;
}
