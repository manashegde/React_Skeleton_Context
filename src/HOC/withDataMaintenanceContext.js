import React from "react";
import { DataMaintContextConsumer } from "contexts/dataMaintenance";

export default function withDataMaintenanceContext(WrappedComponent) {
  class C extends React.Component {
    render() {
      return (
        <DataMaintContextConsumer>
          {dataMaintenanceContext => (
            <WrappedComponent
              {...this.props}
              dataMaintenanceContext={dataMaintenanceContext}
            />
          )}
        </DataMaintContextConsumer>
      );
    }
  }
  C.displayName = `withDataMaintenanceContext(${WrappedComponent.displayName ||
    WrappedComponent.name})`;
  return C;
}
