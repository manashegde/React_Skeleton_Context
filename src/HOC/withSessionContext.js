import React from "react";
import { SessionContextConsumer } from "contexts/session";

export default function withSessionContext(WrappedComponent) {
  class C extends React.Component {
    render() {
      return (
        <SessionContextConsumer>
          {session => (
            <WrappedComponent {...this.props} sessionContext={session} />
          )}
        </SessionContextConsumer>
      );
    }
  }
  C.displayName = `withSessionContext(${WrappedComponent.displayName ||
    WrappedComponent.name})`;
  return C;
}
