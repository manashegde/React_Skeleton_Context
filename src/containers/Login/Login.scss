#Login {
  display: flex;
  flex-direction: column;
  height: 100%;
  overflow: auto;
  .login-container {
    flex-grow: 1;
    input.input {
      border-radius: 1px;
    }
    .login-box {
      margin: 0 auto;
      background: linear-gradient(180deg, #5a6a74 4.97%, #3f525e 31.49%);
      box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
      max-width: 500px;
    }
  }
  .page-code {
    display: flex;
    align-items: center;
    justify-content: flex-end;
    padding: 12px 12px 0px 12px;
  }
}
