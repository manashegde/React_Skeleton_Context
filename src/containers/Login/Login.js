import React from "react";
import intl from "react-intl-universal";
import { setInputValue } from "utils/helpers";
import { urls } from "HOC/withApiService";
import compose from "utils/compose";
import Form from "components/Form";
import Label from "components/Label";
import TextField from "components/TextField";
import Button from "components/Button";
import Image from "components/Image";
import Box from "components/Box";
import SubTitle from "components/SubTitle";

import withSessionContext from "HOC/withSessionContext";
import { routePaths } from "containers/Routes";
import "./Login.css";

class Login extends React.Component {
  state = {
    currentState: "idle",
    isChangePassword: false
  };
  _setState = this.setState;
  usernameEl = React.createRef();
  passwordEl = React.createRef();
  newPasswordEl = React.createRef();
  confirmNewPasswordEl = React.createRef();

  componentWillUnmount() {
    this._setState = null;
  }

  componentDidMount() {
    //todo: remove
    setInputValue(this.usernameEl.current, "xbc");
    setInputValue(this.passwordEl.current, "rSWnq?vXdpseBGu8r&H2&Y8s*6Q$@5Uk)");
    ///////////////////////////////////////////////////
  }

  getXbcToken = async () => {
    const res = await this.props.sessionContext.getXbcState({
      url: urls.postLogin
    });
  };

  callIdentityGateway = async () => {
    const { sessionContext } = this.props;
    const credentials = {
      Username: this.usernameEl.current.value, //xbc
      Password: this.passwordEl.current.value //"rSWnq?vXdpseBGu8r&H2&Y8s*6Q$@5Uk)"
    };
    console.log("--------- creds:", credentials);
    await sessionContext.callIdentityGateway({
      data: credentials
    });
  };

  authenticate = () => {
    console.log("hiii", this.props)
    // this.props.history.push(routePaths.home)
    this._setState &&
      this._setState(
        state => ({
          currentState: "loading"
        }),
        async () => {
          try {
            console.log("------- authenticating");
            await this.callIdentityGateway();
            this.props.history.push(routePaths.home)
            // await this.getXbcToken();
          } catch (e) {
            console.log("------- failed to login:", e);
            this.props.sessionContext.addNotification({ message: e });
            this._setState &&
              this._setState({
                currentState: "idle"
              });
          }
        }
      );
  };

  toggleChangePasswordFields = () => {
    this.setState(state => ({
      isChangePassword: !state.isChangePassword
    }));
  };

  changePassword = async () => {
    console.log("changing password");
  };

  render() {
    const { isChangePassword, currentState } = this.state;

    return (
      <div id="Login">
        <div className="login-container">
          <div
            style={{
              display: "flex",
              justifyContent: "center",
              paddingTop: "5em"
            }}
          >
           
          </div>
          <br />
          <Box className="login-box">
           
           
            <Form style={{ maxWidth: "350px", margin: "0 auto" }}>
              <TextField
                forwardRef={this.usernameEl}
                disabled={currentState === "loading"}
                dataTest="login-username-field"
                label={
                  <Label style={{ color: "white", fontWeight: "400" }}>
                    {intl.get("username")}
                  </Label>
                }
              />
              <TextField
                forwardRef={this.passwordEl}
                type="password"
                disabled={currentState === "loading"}
                dataTest="login-password-field"
                label={
                  <Label style={{ color: "white", fontWeight: "400" }}>
                    {intl.get("password")}
                  </Label>
                }
              />
              {isChangePassword && (
                <React.Fragment>
                  <TextField
                    forwardRef={this.newPasswordEl}
                    type="password"
                    dataTest="login-new-password-field"
                    disabled={currentState === "loading"}
                    label={
                      <Label style={{ color: "white", fontWeight: "400" }}>
                        {intl.get("newPassword")}
                      </Label>
                    }
                  />
                  <TextField
                    forwardRef={this.confirmNewPasswordEl}
                    type="password"
                    disabled={currentState === "loading"}
                    dataTest="login-confirm-new-password-field"
                    label={
                      <Label style={{ color: "white", fontWeight: "400" }}>
                        {intl.get("confirmNewPassword")}
                      </Label>
                    }
                  />
                </React.Fragment>
              )}
              <a
                style={{ color: "white" }}
                onClick={this.toggleChangePasswordFields}
              >
                {intl.get("changePassword")}
              </a>
              <div style={{ display: "flex", justifyContent: "flex-end" }}>
                {isChangePassword ? (
                  <Button
                    className={
                      "is-primary" +
                      (currentState === "loading" ? " is-loading" : "")
                    }
                    onClick={this.changePassword}
                    disabled={currentState === "loading"}
                    dataTest="change-password-btn"
                  >
                    {intl.get("changePassword")}
                  </Button>
                ) : (
                  <Button
                    className={
                      "is-primary" +
                      (currentState === "loading" ? " is-loading" : "")
                    }
                    onClick={this.authenticate}
                    disabled={currentState === "loading"}
                    dataTest="login-btn"
                  >
                    {intl.get("login")}
                  </Button>
                )}
              </div>
            </Form>
          </Box>
        </div>
        <div className="page-code">
          <span>LGOO</span>
        </div>
      </div>
    );
  }
}
export default compose(withSessionContext)(Login);

export { Login };
