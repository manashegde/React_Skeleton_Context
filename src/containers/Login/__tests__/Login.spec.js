import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import { Login } from "containers/Login";
import { urls } from "HOC/withApiService";

const props = {
  apiService: {
    post: jest.fn(() =>
      Promise.resolve({
        jwtToken: "abs",
        firstName: "first",
        lastName: "last",
        roles: ["admin"]
      })
    ),
    urls
  },
  sessionContext: {
    setAuth: jest.fn(),
    addNotification: jest.fn()
  }
};

const setup = () => {
  const wrapper = mount(<Login {...props} />);
  const instance = wrapper.instance();
  return {
    wrapper,
    instance
  };
};

describe("Login", () => {
  it("should match snapshot", () => {
    const tree = toJson(mount(<Login />));
    expect(tree).toMatchSnapshot();
  });

  it("should call apiService POST to login with username and password successfully", async () => {
    const { wrapper, instance } = setup();

    const usernameField = wrapper.find('[data-test="login-username-field"]');
    const passwordField = wrapper.find('[data-test="login-password-field"]');

    usernameField.simulate("change", { target: { value: "user" } });
    passwordField.simulate("change", { target: { value: "pass" } });

    await instance.login();
    expect(props.apiService.post).toHaveBeenCalled();
    expect(props.apiService.post).toHaveBeenCalledWith({
      url: props.apiService.urls.login,
      data: {
        username: "user",
        password: "pass"
      }
    });
    expect(props.sessionContext.setAuth).toHaveBeenCalled();
    expect(props.sessionContext.addNotification).not.toHaveBeenCalled();
  });

  it("should show a notification if there was an error signing in", async () => {
    props.apiService.post = jest.fn(() => Promise.reject("failed to login"));
    const { wrapper, instance } = setup();
    await instance.login();
    expect(props.sessionContext.addNotification).toHaveBeenCalled();
  });
});
