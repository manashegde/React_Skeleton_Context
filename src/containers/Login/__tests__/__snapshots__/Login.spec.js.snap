// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match snapshot 1`] = `
<Login>
  <div
    id="Login"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "height": "100%",
        "overflow": "auto",
        "paddingTop": "5em",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      <Image
        src="logo-main.png"
      >
        <img
          alt="blank"
          src="logo-main.png"
        />
      </Image>
      <Image
        src="logo-secondary.png"
        style={
          Object {
            "width": "260px",
          }
        }
      >
        <img
          alt="blank"
          src="logo-secondary.png"
          style={
            Object {
              "width": "260px",
            }
          }
        />
      </Image>
    </div>
    <br />
    <Box
      style={
        Object {
          "background": "linear-gradient(180deg, #5A6A74 4.97%, #3F525E 31.49%)",
          "boxShadow": "0px 4px 4px rgba(0, 0, 0, 0.25)",
        }
      }
    >
      <div
        className="box"
        style={
          Object {
            "background": "linear-gradient(180deg, #5A6A74 4.97%, #3F525E 31.49%)",
            "boxShadow": "0px 4px 4px rgba(0, 0, 0, 0.25)",
          }
        }
      >
        <SubTitle
          is="4"
          style={
            Object {
              "color": "white",
              "textAlign": "center",
            }
          }
        >
          <h5
            className="subtitle is-4"
            style={
              Object {
                "color": "white",
                "textAlign": "center",
              }
            }
          />
        </SubTitle>
        <SubTitle
          is="4"
          style={
            Object {
              "color": "white",
              "textAlign": "center",
            }
          }
        >
          <h5
            className="subtitle is-4"
            style={
              Object {
                "color": "white",
                "textAlign": "center",
              }
            }
          />
        </SubTitle>
        <Form
          style={
            Object {
              "margin": "0 auto",
              "maxWidth": "350px",
            }
          }
        >
          <form
            noValidate={true}
            onSubmit={[Function]}
            style={
              Object {
                "margin": "0 auto",
                "maxWidth": "350px",
              }
            }
          >
            <TextField
              dataTest="login-username-field"
              disabled={false}
              label={[Function]}
            >
              <div
                className="field"
              >
                <Label
                  style={
                    Object {
                      "color": "white",
                      "fontWeight": "400",
                    }
                  }
                >
                  <label
                    className="label"
                    style={
                      Object {
                        "color": "white",
                        "fontWeight": "400",
                      }
                    }
                  />
                </Label>
                <div
                  className="control"
                >
                  <input
                    className="input"
                    data-test="login-username-field"
                    disabled={false}
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                  <p
                    className="help is-warning"
                    style={
                      Object {
                        "fontSize": "15px",
                      }
                    }
                  />
                </div>
              </div>
            </TextField>
            <TextField
              dataTest="login-password-field"
              disabled={false}
              label={[Function]}
              type="password"
            >
              <div
                className="field"
              >
                <Label
                  style={
                    Object {
                      "color": "white",
                      "fontWeight": "400",
                    }
                  }
                >
                  <label
                    className="label"
                    style={
                      Object {
                        "color": "white",
                        "fontWeight": "400",
                      }
                    }
                  />
                </Label>
                <div
                  className="control"
                >
                  <input
                    className="input"
                    data-test="login-password-field"
                    disabled={false}
                    onChange={[Function]}
                    type="password"
                    value=""
                  />
                  <p
                    className="help is-warning"
                    style={
                      Object {
                        "fontSize": "15px",
                      }
                    }
                  />
                </div>
              </div>
            </TextField>
            <div
              style={
                Object {
                  "textAlign": "center",
                }
              }
            >
              <a
                style={
                  Object {
                    "color": "white",
                  }
                }
              />
            </div>
            <div
              style={
                Object {
                  "display": "flex",
                  "justifyContent": "flex-end",
                }
              }
            >
              <Button
                className="is-primary"
                dataTest="login-btn"
                disabled={false}
                onClick={[Function]}
              >
                <button
                  className="button is-primary"
                  data-test="login-btn"
                  disabled={false}
                  onClick={[Function]}
                />
              </Button>
            </div>
          </form>
        </Form>
      </div>
    </Box>
  </div>
</Login>
`;
